File: backend/ustat_engine/stats/ttest.py
"""Independent-samples and one-sample t-test.

Extracted from the `POST /api/stats/ttest` handler in
routers/stats/inferential.py. The computational body is verbatim: both
branches, every expression and every string, are the ones that were there
before. That is the whole point -- this module has to be provably the same
arithmetic as the code it replaced, now runnable in either runtime.

First analysis to move that actually reads a dataset. Power and meta-analysis
compute from numbers typed into a form; this one needs rows, so it is the first
real exercise of the frame path: `build_envelope` on the server,
`frame_from_envelope` in the worker, and `registry.run` refusing the frame if
the Select Cases it was cut under has moved on.

Translation from the FastAPI original, and nothing else:
  - `class TTestRequest(BaseModel)` became the `_P` shim below, with the same
    field defaults and the same `group_col` alias, so the copied body still
    says `req.<field>`. Keeping those references intact is what lets this file
    be diffed against its source.
  - `_get_df(req.session_id)` disappeared: the frame is a parameter now. The
    session lookup is the server's business and stayed in the router.
  - `raise HTTPException(status_code=X, detail=Y)` became
    `raise EngineError(Y, status_hint=X)`, detail strings unchanged.
  - `_sanitize(ret)` became `jsonsafe.sanitize(ret)` -- the same recursion over
    the same non-finite floats, plus numpy scalars the original left to
    FastAPI's encoder. The browser has no such encoder behind it.

pandas and `clean_two_level` are imported inside `_two_level_work` rather than
at module scope. `ustat_engine/__init__.py` imports this module so the registry
is complete before anything asks it a question, and a module-level `import
pandas` would therefore make `import ustat_engine` fail in a browser that had
only loaded numpy/scipy/statsmodels for a power analysis.
"""
from __future__ import annotations

from scipy import stats as scipy_stats

from ..errors import EngineError
from ..jsonsafe import sanitize
from ..registry import register
from ..spec import AnalysisSpec
from ..text.ttest import (
    methods_ttest_ind, methods_ttest_one,
    results_ttest_ind, results_ttest_one,
    r_ttest_ind, r_ttest_one,
)
from .assumptions import check_equal_variances, check_normality
from .effect_sizes import (
    cohen_d, cohen_d_one_sample, group_summary, welch_satterthwaite_df,
)

_METHODS = ("auto", "student", "welch")


class _P:
    """Shim standing in for the pydantic `TTestRequest`, so the copied
    function body below can keep using `req.<field>` unchanged. Defaults
    mirror the field declarations of `TTestRequest` exactly, including the
    `group_col` validation alias."""

    def __init__(self, params):
        if not params.get("column"):
            raise EngineError("Field 'column' is required.", status_hint=422)
        self.column = params["column"]
        # AliasChoices("group_column", "group_col") — either spelling arrives
        # from a caller that predates the rename.
        self.group_column = params.get("group_column")
        if self.group_column is None:
            self.group_column = params.get("group_col")
        mu = params.get("mu", 0.0)
        self.mu = 0.0 if mu is None else mu
        # "auto" lets Levene pick Student vs Welch; the other two force it.
        self.method = params.get("method") or "auto"
        if self.method not in _METHODS:
            raise EngineError(
                f"Field 'method' must be one of {_METHODS}.", status_hint=422
            )
        # Legacy alias, kept so existing callers keep working. None means "not
        # supplied" — it used to default to True while the handler ignored it
        # entirely and always let Levene decide.
        self.equal_var = params.get("equal_var")


def _two_level_work(df, value_col: str, group_col: str):
    import pandas as pd

    from ..frame.category_health import clean_two_level

    work = df[[value_col, group_col]].copy()
    cleaned = clean_two_level(work[group_col])
    work[group_col] = cleaned.series
    work[value_col] = pd.to_numeric(work[value_col], errors="coerce")
    return work.dropna(), cleaned.warnings


# ─────────────────────────────────────────────────────────────────────────
# Everything below this line is copied verbatim (HTTPException raises
# rewritten to EngineError, `_sanitize` to `sanitize`, `df` now a parameter
# instead of a session lookup, and `req` bound to the `_P` shim instead of a
# pydantic model) from the `ttest` handler in
# backend/routers/stats/inferential.py.
# ─────────────────────────────────────────────────────────────────────────

def run_ttest(frame, params: dict) -> dict:
    # Deferred for the same reason as pandas in `_two_level_work`:
    # `frame.levels` imports pandas at module scope, and this module is
    # imported at `import ustat_engine` time.
    from ..frame.levels import sorted_groups

    req = _P(params)
    df = frame
    col = df[req.column].dropna()

    if req.group_column:
        work, warnings = _two_level_work(df, req.column, req.group_column)
        groups = sorted_groups(work[req.group_column])
        if len(groups) != 2:
            raise EngineError("Group column must have exactly 2 groups", status_hint=400)
        g1 = work[work[req.group_column] == groups[0]][req.column].values.astype(float)
        g2 = work[work[req.group_column] == groups[1]][req.column].values.astype(float)

        # Assumption checks
        assumptions = [check_normality(g1, str(groups[0])), check_normality(g2, str(groups[1])),
                       check_equal_variances([g1, g2], [str(groups[0]), str(groups[1])],
                                             on_violation="Welch correction applied")]
        # Precedence: explicit method > legacy equal_var > Levene.
        if req.method == "welch":
            use_welch, chosen_by = True, "request (method)"
        elif req.method == "student":
            use_welch, chosen_by = False, "request (method)"
        elif req.equal_var is not None:
            use_welch, chosen_by = (not req.equal_var), "request (equal_var)"
        else:
            use_welch, chosen_by = (not assumptions[2]["met"]), "auto (Levene)"
        stat, p = scipy_stats.ttest_ind(g1, g2, equal_var=not use_welch)
        sig = bool(p < 0.05)
        es = cohen_d(g1, g2)
        p_str = '<0.001' if p < 0.001 else f'{p:.4f}'

        ret = {
            "test": f"Independent samples t-test{' (Welch)' if use_welch else ''}",
            "group1": str(groups[0]), "n1": len(g1), "mean1": float(g1.mean()),
            "group2": str(groups[1]), "n2": len(g2), "mean2": float(g2.mean()),
            # df must match the test that produced t and p. Welch uses the
            # fractional Satterthwaite df; only the pooled test uses n1+n2-2.
            "t": float(stat), "p": float(p),
            "df": welch_satterthwaite_df(g1, g2) if use_welch else float(len(g1) + len(g2) - 2),
            "df_method": "welch_satterthwaite" if use_welch else "pooled",
            "variance_assumption": "welch" if use_welch else "student",
            "variance_assumption_selected_by": chosen_by,
            "significant": sig,
            "effect_sizes": [es],
            "assumptions": assumptions,
            "summary": {str(groups[0]): group_summary(g1, str(groups[0])),
                        str(groups[1]): group_summary(g2, str(groups[1]))},
            "interpretation": f"{'Significant' if sig else 'No significant'} difference between groups (t = {stat:.3f}, p = {p_str}, Hedges' g = {es['value']:.3f} [{es['magnitude']}])",
            "methods_text": methods_ttest_ind(req.column, req.group_column, use_welch),
            "r_code": r_ttest_ind(req.column, req.group_column, use_welch),
        }
        if warnings:
            ret["warnings"] = warnings
        ret["result_text"] = results_ttest_ind(ret)
        return sanitize(ret)
    else:
        x = col.astype(float).values
        stat, p = scipy_stats.ttest_1samp(x, req.mu)
        sig = bool(p < 0.05)
        es = cohen_d_one_sample(x, req.mu)
        p_str = '<0.001' if p < 0.001 else f'{p:.4f}'

        ret = {
            "test": "One-sample t-test",
            "mu": req.mu, "n": len(x),
            "mean": float(x.mean()), "std": float(x.std(ddof=1)),
            "t": float(stat), "p": float(p), "df": int(len(x) - 1),
            "significant": sig,
            "effect_sizes": [es],
            "assumptions": [check_normality(x, req.column)],
            "summary": {"sample": group_summary(x, "Sample")},
            "interpretation": f"Mean {'differs from' if sig else 'does not differ from'} {req.mu} (t = {stat:.3f}, p = {p_str}, Cohen's d = {es['value']:.3f} [{es['magnitude']}])",
            "methods_text": methods_ttest_one(req.column, req.mu),
            "r_code": r_ttest_one(req.column, req.mu),
        }
        ret["result_text"] = results_ttest_one(ret)
        return sanitize(ret)


register(AnalysisSpec(
    id="stats.ttest",
    fn=run_ttest,
    needs_frame=True,
    # pandas because the frame arrives as one; scipy for the test itself and
    # for Shapiro/Levene.
    #
    # statsmodels is here reluctantly, and it is the one place this spec is not
    # the obvious minimum. `check_normality` switches to the Lilliefors test at
    # n >= 50 and imports statsmodels *inside the function* to do it -- which on
    # the server is a lazy import and in the browser is a hard failure, because
    # a Pyodide package that was not in the boot plan cannot be fetched from
    # inside a synchronous `runPython`. Any group of 50 or more would crash a
    # local t-test that declared only numpy/scipy/pandas, so the dependency is
    # declared for the branch rather than for the common case.
    deps=("numpy", "scipy", "pandas", "statsmodels"),
    required_columns=lambda p: [
        p.get("column"),
        p.get("group_column") or p.get("group_col"),
    ],
    cost_key="stats.ttest",
    doc=(
        "Independent-samples t-test (Student or Welch, chosen by Levene unless "
        "the caller forces one) or a one-sample t-test against mu."
    ),
    tags=("inferential", "parametric"),
))


__all__ = ["run_ttest"]

File: backend/ustat_engine_r/analyses/ttest.R
# Independent-samples and one-sample t-test.
#
# The R half of ustat_engine/stats/ttest.py, and deliberately the thinnest file
# in the tree. Every number it returns comes out of stats::t.test, out of the
# helpers in runtime/stats.R, or out of the frame -- there is no arithmetic
# here. That is what "two engines, both thin" means: this file decides which
# test to run, on which rows, and what to call the answer; R decides what the
# answer is.
#
# WHAT MUST MATCH, AND WHAT DOES NOT
# -----------------------------------
# The KEYS of the returned list are a contract: HypothesisPanel destructures
# them with no optional chaining, so a missing one is a runtime crash rather
# than a type error, and the browser has to be able to show an R result in the
# panel a Python result came out of. test_r_ttest_local.py asserts set equality
# against run_ttest's own keys for the same params.
#
# The VALUES match to about 1e-15 for everything that comes from t.test, and
# exactly for the counts, the means and the rounded effect sizes. Two do not,
# both recorded in qa/parity/ttest.json:
#
#   - the Lilliefors p in the normality assumption (different approximation of
#     the same null distribution -- see runtime/stats.R);
#   - the wording around `mu` when a caller sends it as a JSON integer (see
#     ustat_py_num_text in runtime/text.R).
#
# PRECEDENCE for Student vs Welch is copied exactly: an explicit `method` wins,
# then the legacy `equal_var`, then Levene. Getting that order wrong would not
# show up as an error, it would silently run the other test.

USTAT_TTEST_METHODS <- c("auto", "student", "welch")

ustat_ttest_params <- function(params) {
  column <- params$column
  if (is.null(column) || length(column) != 1L || is.na(column) ||
      !nzchar(as.character(column))) {
    ustat_stop("Field 'column' is required.", 422L)
  }

  # AliasChoices("group_column", "group_col") -- either spelling arrives from a
  # caller that predates the rename.
  group_column <- params$group_column
  if (is.null(group_column)) {
    group_column <- params$group_col
  }
  if (!is.null(group_column) &&
      (is.na(group_column) || !nzchar(as.character(group_column)))) {
    group_column <- NULL
  }

  method <- params$method
  if (is.null(method) || is.na(method) || !nzchar(as.character(method))) {
    method <- "auto"
  }
  method <- as.character(method)
  if (!(method %in% USTAT_TTEST_METHODS)) {
    ustat_stop(
      "Field 'method' must be one of ('auto', 'student', 'welch').", 422L
    )
  }

  # mu defaults to the float 0.0 in the Python request shim, and NULL means "not
  # supplied" for equal_var -- it used to default to TRUE while the handler
  # ignored it entirely and always let Levene decide.
  mu_supplied <- !is.null(params$mu)
  list(
    column = as.character(column),
    group_column = if (is.null(group_column)) NULL else as.character(group_column),
    method = method,
    equal_var = params$equal_var,
    mu = if (mu_supplied) as.numeric(params$mu) else 0,
    mu_text = ustat_py_num_text(
      if (mu_supplied) as.numeric(params$mu) else 0,
      as_float = !mu_supplied
    )
  )
}

ustat_ttest_column <- function(frame, name) {
  if (!(name %in% names(frame))) {
    ustat_stop(paste0("Unknown column: '", name, "'"), 400L)
  }
  frame[[name]]
}

ustat_ttest_ind <- function(req, frame) {
  # _two_level_work: normalise the grouping column, coerce the value column,
  # then drop any row missing either. This is what fixes n1 and n2.
  cleaned <- ustat_clean_two_level(
    ustat_ttest_column(frame, req$group_column), name = req$group_column
  )
  value <- ustat_to_numeric(ustat_ttest_column(frame, req$column))
  grp <- cleaned$series
  keep <- !is.na(value) & !is.na(grp)
  value <- value[keep]
  grp <- grp[keep]

  groups <- ustat_sorted_groups(grp)
  if (length(groups) != 2L) {
    ustat_stop("Group column must have exactly 2 groups", 400L)
  }
  g1 <- value[grp == groups[1]]
  g2 <- value[grp == groups[2]]
  name1 <- as.character(groups[1])
  name2 <- as.character(groups[2])

  assumptions <- list(
    ustat_check_normality(g1, name1),
    ustat_check_normality(g2, name2),
    ustat_check_equal_variances(
      list(g1, g2), c(name1, name2), "Welch correction applied"
    )
  )

  if (identical(req$method, "welch")) {
    use_welch <- TRUE
    chosen_by <- "request (method)"
  } else if (identical(req$method, "student")) {
    use_welch <- FALSE
    chosen_by <- "request (method)"
  } else if (!is.null(req$equal_var)) {
    use_welch <- !isTRUE(as.logical(req$equal_var))
    chosen_by <- "request (equal_var)"
  } else {
    use_welch <- !isTRUE(assumptions[[3]]$met)
    chosen_by <- "auto (Levene)"
  }

  tt <- stats::t.test(g1, g2, var.equal = !use_welch)
  stat <- unname(tt$statistic)
  p <- unname(tt$p.value)
  sig <- isTRUE(p < 0.05)
  es <- ustat_cohen_d(g1, g2)

  ret <- list(
    test = paste0(
      "Independent samples t-test", if (use_welch) " (Welch)" else ""
    ),
    group1 = name1, n1 = length(g1), mean1 = mean(g1),
    group2 = name2, n2 = length(g2), mean2 = mean(g2),
    t = stat,
    p = p,
    # df must match the test that produced t and p. t.test reports the
    # fractional Satterthwaite df for Welch and n1 + n2 - 2 for the pooled
    # test, which is exactly the pairing the Python side assembles by hand.
    df = unname(tt$parameter),
    df_method = if (use_welch) "welch_satterthwaite" else "pooled",
    variance_assumption = if (use_welch) "welch" else "student",
    variance_assumption_selected_by = chosen_by,
    significant = sig,
    effect_sizes = list(es),
    assumptions = assumptions,
    summary = stats::setNames(
      list(ustat_group_summary(g1, name1), ustat_group_summary(g2, name2)),
      c(name1, name2)
    ),
    interpretation = ustat_interpretation_ttest_ind_text(sig, stat, p, es),
    methods_text = ustat_methods_ttest_ind_text(
      req$column, req$group_column, use_welch
    ),
    r_code = ustat_r_ttest_ind_text(req$column, req$group_column, use_welch)
  )
  if (length(cleaned$warnings) > 0L) {
    ret$warnings <- cleaned$warnings
  }
  ret$result_text <- ustat_results_ttest_ind_text(ret)
  ret
}

ustat_ttest_one <- function(req, frame) {
  x <- ustat_to_numeric(ustat_ttest_column(frame, req$column))
  x <- x[!is.na(x)]

  tt <- stats::t.test(x, mu = req$mu)
  stat <- unname(tt$statistic)
  p <- unname(tt$p.value)
  sig <- isTRUE(p < 0.05)
  es <- ustat_cohen_d_one_sample(x, req$mu)

  ret <- list(
    test = "One-sample t-test",
    mu = req$mu, n = length(x),
    mean = mean(x), std = stats::sd(x),
    t = stat, p = p, df = length(x) - 1L,
    significant = sig,
    effect_sizes = list(es),
    assumptions = list(ustat_check_normality(x, req$column)),
    summary = list(sample = ustat_group_summary(x, "Sample")),
    interpretation = ustat_interpretation_ttest_one_text(
      sig, stat, p, es, req$mu_text
    ),
    methods_text = ustat_methods_ttest_one_text(req$column, req$mu_text),
    r_code = ustat_r_ttest_one_text(req$column, req$mu_text)
  )
  # mu_text is passed rather than stored: it is a rendering of a field the
  # payload already carries, and a second spelling of mu in the result would be
  # one more thing the two engines could disagree about.
  ret$result_text <- ustat_results_ttest_one_text(
    c(ret, list(mu_text = req$mu_text))
  )
  ret
}

ustat_ttest <- function(params, frame) {
  req <- ustat_ttest_params(params)
  if (is.null(req$group_column)) {
    ustat_ttest_one(req, frame)
  } else {
    ustat_ttest_ind(req, frame)
  }
}

ustat_register(list(
  id = "stats.ttest",
  needs_frame = TRUE,
  # Base R carries t.test, shapiro.test and the one-way F this engine's Levene
  # goes through. These two are here for one branch each of check_normality, and
  # for the same reason the Python spec declares statsmodels reluctantly: a
  # package that is not in the boot plan cannot be fetched from inside a
  # synchronous call, so any group of 50 or more would crash a local run that
  # declared only base R.
  packages = c("moments", "nortest"),
  columns_for = function(params) {
    group_column <- params$group_column
    if (is.null(group_column)) group_column <- params$group_col
    c(params$column, group_column)
  },
  fn = ustat_ttest
))
